File: Language/Go/design-patterns/behavioural-patterns/observer/src/publisher.go
package src

import (
	"fmt"
	"time"
)

type Publisher struct {
	subscriptions []func(string)
}

func (p *Publisher) Start() {
	fmt.Println("Publisher: Wasting time...")
	time.Sleep(time.Second * 5)
	p.end()
}

func (p *Publisher) end() {
	fmt.Println("Publisher: Notifying subscribers that I am done wasting time...")
	p.notify("Done")
}

func (p *Publisher) notify(d string) {
	for _, s := range p.subscriptions {
		s(d)
	}
}

func (p *Publisher) Subscribe(f func(string)) int {
	fmt.Println("Publisher: Subscribing a subscriber...")

	p.subscriptions = append(p.subscriptions, f)

	return len(p.subscriptions) - 1
}

func (p *Publisher) Unsubscribe(i int) {
	fmt.Println("Publisher: Unsubscribing subscriber...")

	p.subscriptions = append(p.subscriptions[:i], p.subscriptions[i+1:len(p.subscriptions)]...)
}

File: Language/Scala/design-patterns/behavioural-patterns/observer/src/main/scala/Publisher.scala
package observer

class Publisher {
  var subscriptions: List[String => Unit] = List()

  def start(): Unit = {
    println("Publisher: Wasting time...")
    Thread.sleep(5000)
    end()
  }

  def end(): Unit = {
    println("Publisher: Notifying subscribers that I am done wasting time...")
    notify("Done")
  }

  def notify(d: String): Unit = {
    subscriptions.foreach(_(d))
  }

  def subscribe(f: String => Unit): Option[Int] = {
    println("Publisher: Subscribing a subscriber...")
    subscriptions = subscriptions ++ List(f)
    Some(subscriptions.length - 1)
  }

  def unsubscribe(i: Option[Int]): Unit = {
    println("Publisher: Unsubscribing a subscriber...")
    i match {
      case Some(x) =>
        subscriptions = subscriptions.patch(x, Nil, 1)
        println(s"Publisher: removing subscriber $x")
      case None =>
    }
  }
}
